components/layout_2020/table/layout.rs

```diff
@@ -918,88 +1007,356 @@ impl<'a> TableLayout<'a> {
+            if a.range() == b.range() {
+                return a
+                    .measure
+                    .content_sizes
+                    .min_content
+                    .cmp(&b.measure.content_sizes.min_content);
+            }
+            if a.fully_encloses(b) {
+                return std::cmp::Ordering::Greater;
+            }
+            if b.fully_encloses(a) {
+                return std::cmp::Ordering::Less;
+            }
+            a.coordinates.y.cmp(&b.coordinates.y)
+        });
+
+        for rowspan_to_distribute in cells_to_distribute {
+            let rows_spanned = rowspan_to_distribute.range();
+            let current_rows_size: Au = rows_spanned.clone().map(|index| row_sizes[index]).sum();
+            let border_spacing_spanned =
+                self.table.border_spacing().block * (rows_spanned.len() - 1) as i32;
+            let excess_size = (rowspan_to_distribute.measure.content_sizes.min_content -
+                current_rows_size -
+                border_spacing_spanned)
+                .max(Au::zero());
+
+            self.distribute_extra_size_to_rows(
+                excess_size,
+                rows_spanned,
+                row_sizes,
+                None,
+                true, /* rowspan_distribution */
+            );
+        }
+    }
+
+    /// An implementation of the same extra block size distribution algorithm used in
+    /// LayoutNG and described at <https://github.com/w3c/csswg-drafts/issues/4418>.
+    fn distribute_extra_size_to_rows(
+        &self,
+        mut excess_size: Au,
+        track_range: Range<usize>,
+        track_sizes: &mut Vec<Au>,
+        percentage_resolution_size: Option<Au>,
+        rowspan_distribution: bool,
+    ) {
+        if excess_size.is_zero() {
+            return;
+        }
+
+        let is_constrained = |track_index: &usize| self.rows[*track_index].constrained;
+        let is_unconstrained = |track_index: &usize| !is_constrained(track_index);
+        let is_empty: Vec<bool> = track_sizes.iter().map(|size| size.is_zero()).collect();
+        let is_not_empty = |track_index: &usize| !is_empty[*track_index];
+        let other_row_that_starts_a_rowspan = |track_index: &usize| {
+            *track_index != track_range.start &&
+                self.rows[*track_index].has_cell_with_span_greater_than_one
+        };
+
+        // If we have a table height (not during rowspan distribution), first distribute to rows
+        // that have percentage sizes proportionally to the size missing to reach the percentage
+        // of table height required.
+        if let Some(percentage_resolution_size) = percentage_resolution_size {
+            let get_percent_block_size_deficit = |row_index: usize, track_size: Au| {
+                let size_needed_for_percent = percentage_resolution_size
+                    .scale_by(self.rows[row_index].percent.0 as f32 / 100.);
+                size_needed_for_percent - track_size.max(Au::zero())
```

Comment:
Should this be?

```suggestion
                (size_needed_for_percent - track_size).max(Au::zero())
```

Or possibly leave it to the callers, since one is already checking `> Au::zero()` and doesn't need `.max(Au::zero())`

Reply:
I fixed this one. I think we do need this in case the sum of the percentages of the rows is greater than 100%.